pkg/stores/proxy/cache/cache.go

```diff
@@ -0,0 +1,150 @@
+// Package cache implements the sized revision cache. The size revision cache maintains a cache of Unstructuredlists while
+// attempting to limit the aggregate values stored by the given memory size (in bytes) parameter.
+package cache
+
+import (
+	"fmt"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/cache"
+)
+
+var (
+	ErrNotFound = errors.New("key was not found in cache")
+)
+
+// SizedRevisionCache maintains a cache of Unstructuredlists while attempting to limit the aggregate values stored by
+// the given memory size (in bytes) parameter.
+type SizedRevisionCache struct {
+	listRevisionCache *cache.LRUExpireCache
+	cacheLock         sync.Mutex
+	size              int
+	sizeLimit         int
+}
+
+// Cacher is a cache that stores and retrieves UnstructuredLists.
+type Cacher interface {
+	Get(key CacheKey) (*unstructured.UnstructuredList, error)
+	Add(key CacheKey, list *unstructured.UnstructuredList) error
+}
+
+type CacheKey struct {
+	listOptions  v1.ListOptions
+	resourcePath string
+	namespace    string
+	cont         string
+}
+
+type cacheObj struct {
+	size int
+	obj  interface{}
+}
+
+// NewSizedRevisionCache accepts a sizeLimit, in bytes, and a maxElements parameter to create a SizedRevisionCache.
+func NewSizedRevisionCache(sizeLimit, maxElements int) *SizedRevisionCache {
+	return &SizedRevisionCache{
+		listRevisionCache: cache.NewLRUExpireCache(maxElements),
+		sizeLimit:         sizeLimit,
+	}
+}
+
+// String returns a string contains the fields values of the cacheKey receiver.
+func (c CacheKey) String() string {
+	return fmt.Sprintf("listOptions: %v, resourcePath: %s, namespace: %s, cont: %s", c.listOptions.String(), c.resourcePath, c.namespace, c.cont)
+}
+
+// Get returns the UnstructuredList stored under the given cacheKey if available. If not, as error is returned.
+func (s *SizedRevisionCache) Get(key CacheKey) (*unstructured.UnstructuredList, error) {
+	s.cacheLock.Lock()
+	defer s.cacheLock.Unlock()
+	// check if cache stored all namespaces
+
+	obj, ok := s.listRevisionCache.Get(key)
+	if !ok {
+		return nil, ErrNotFound
+	}
+	uList, ok := obj.(cacheObj)
+	if !ok {
+		return nil, fmt.Errorf("could not assert object stored with key [%s] key as UnstructuredList", key)
+	}
+	return uList.obj.(*unstructured.UnstructuredList), nil
+}
+
+// Add attempts to add the given UnstructuredListed under the given key.
+func (s *SizedRevisionCache) Add(key CacheKey, list *unstructured.UnstructuredList) error {
+	s.cacheLock.Lock()
+	defer s.cacheLock.Unlock()
+
+	objBytes, err := list.MarshalJSON()
```

Comment:
Is my understanding correct that intent here is to marshal `list` to JSON only to determine its cache size?

If so, I would suggest trying to see if there are any other ways to determine memory consumption, because:
 * JSON representation consumes by construction many more bytes than raw objects, so the estimation might not be accurate
 * JSON marshaling can be expected to be CPU heavy, especially on large lists of complicated objects, and
 * conversion would generate a significant amount of garbage to be collected (generally speaking `objBytes`, for each returned resource list)

Reply:
Could use https://pkg.go.dev/unsafe?#Sizeof or https://pkg.go.dev/reflect#Type `.Size()` maybe?

Reply:
Unfortunately, I don't think Sizeof will work because we are working with slices unless we use size of on every element: `Sizeof takes an expression x of any type and returns the size in bytes of a hypothetical variable v as if v was declared via var v = x` and Size would be the same:`it is analogous to unsafe.Sizeof`. It doesn't sound efficient but we could bench mark it anyway to make sure.

I do think one relief here is that this only happens when adding an object. It doesn't happen when reevaluating the aggregate size since we store the value. I think if this proves to increase the response time drastically, we can add it to the cache in a goroutine, I just did not want to do that preemptively. I meant to add this to the `Ways this approach could be improved:` section. I will do that now.

Reply:
Iterating through the list and calling `Sizeof` (recursively, if necessary) sounds much better than serializing the collection to JSON to me.